db/compaction/compaction_job.cc

```diff
@@ -290,6 +296,128 @@ void CompactionJob::Prepare() {
   }
 }
 
+uint64_t CompactionJob::GetNumPlannedSubcompactions(
+    uint64_t max_output_files_limit, uint64_t ranges_size_limit) {
+  auto* c = compact_->compaction;
+  uint64_t max_subcompactions_limit =
+      static_cast<uint64_t>(c->max_subcompactions());
+  uint64_t num_planned_subcompactions =
+      std::min({static_cast<uint64_t>(ranges_size_limit),
+                max_subcompactions_limit, max_output_files_limit});
+  if (c->immutable_options()->compaction_pri == kRoundRobin) {
+    // For round-robin compaction prioity, we may have a large compaction with
+    // multiple input files in the start level. In this case, we need to employ
+    // more subcompactions (may exceed the max_subcompaction limit). The extra
+    // subcompactions will be executed using reserved threads and taken into
+    // account bg_compaction_scheduled. The number of planned subcompaction is
+    // determined through 3 steps:
+
+    // Step 1: Initialized against max_output_file_limit, the number of input
+    // files in the start level, and also the range size limit
+    num_planned_subcompactions = std::min(
+        {static_cast<uint64_t>(c->num_input_files(0)), max_output_files_limit,
+         static_cast<uint64_t>(ranges_size_limit)});
```

Comment:
Can we put these into a function like `num_planned_subcompactions = CompactionJob::InitializeNumPlannedSubcompactions(max_output_files_limit, ranges_size_limit){
if (c->immutable_options()->compaction_pri == kRoundRobin){

} else {

}
}`

For the rest of the function `CompactionJob::GetNumPlannedSubcompactions`, we can put them into `num_planned_subcompactions = ReserveSubcompactionResources(num_planned_subcompactions){
if (c->immutable_options()->compaction_pri == kRoundRobin){
  // Update num_planned_subcompactions based on per-db constraint and env threads
} else {
  return num_planned_subcompactions;
}
}`

Reply:
Your design looks much cleaner. Will work on this.